webapp/packages/core-executor/src/whileTask.ts

```diff
@@ -6,35 +6,54 @@
  * you may not use this file except in compliance with the License.
  */
 import { AutoRunningTask } from './TaskScheduler/AutoRunningTask';
+import { CancelError } from './TaskScheduler/CancelError';
 import type { ITask } from './TaskScheduler/ITask';
 import { Task } from './TaskScheduler/Task';
+import { TimeoutError } from './TaskScheduler/TimeoutError';
 
 export function whileTask<T>(
   callback: (value: T) => Promise<boolean> | boolean,
   task: () => Promise<T>,
   interval: number,
   cancelMessage?: string,
+  timeout?: number,
 ): ITask<T> {
   let resolve: (value: T | PromiseLike<T>) => void;
   let reject: (reason?: any) => void;
+  let fulfilled = false;
 
-  const lockPromise = new Promise<T>((_resolve, _reject) => {
-    resolve = _resolve;
-    reject = _reject;
-  });
-
+  let intervalTimeoutId: NodeJS.Timeout | null;
   let timeoutId: NodeJS.Timeout | null;
   let activeTask: Promise<T> | null;
   let stopped = false;
 
-  function stop() {
+  const lockPromise = new Promise<T>((_resolve, _reject) => {
+    resolve = _resolve;
+    reject = _reject;
+  }).finally(() => {
+    if (intervalTimeoutId) {
+      clearTimeout(intervalTimeoutId);
+    }
+
     if (timeoutId) {
```

Comment:
what if we want to use this with `timeout = 0`?
I suppose it is better to use here:
```suggestion
    if (isNotNullDefined(timeoutId)) {
```

Reply:
please check all `if` statements for this

Reply:
fixed